File: phase2_web_application/js/algorithms.js
// ============================================================================
// 核心算法类区域 (CORE ALGORITHM CLASSES)
// ============================================================================
// 包含角度平滑器、滞后阈值评估器、自适应阈值管理器等核心算法类
// ============================================================================

// ========== Phase 1 优化：加权移动平均平滑器（WMA）==========
// 改进：更重视最近的数据，响应更快，平滑效果更好
class AngleSmoother {
    constructor(windowSize) {
        // 如果 windowSize 未提供，使用默认值 10
        if (windowSize === undefined) {
            windowSize = typeof SMOOTHING_WINDOW_SIZE !== 'undefined' ? SMOOTHING_WINDOW_SIZE : 10;
        }
        this.windowSize = windowSize;
        this.neckHistory = [];
        this.torsoHistory = [];
    }
    
    /**
     * 生成线性权重（最近的数据权重更大）
     */
    generateWeights(length) {
        const weights = [];
        for (let i = 0; i < length; i++) {
            weights.push(i + 1); // 权重递增：1, 2, 3, ...
        }
        const sum = weights.reduce((a, b) => a + b, 0);
        return weights.map(w => w / sum); // 归一化
    }
    
    /**
     * 加权平均计算
     */
    weightedAverage(values, weights) {
        return values.reduce((sum, val, i) => sum + val * weights[i], 0);
    }
    
    smooth(neckAngle, torsoAngle) {
        // 添加新值
        this.neckHistory.push(neckAngle);
        this.torsoHistory.push(torsoAngle);
        
        // 保持窗口大小
        if (this.neckHistory.length > this.windowSize) {
            this.neckHistory.shift();
            this.torsoHistory.shift();
        }
        
        // 加权移动平均（WMA）：更重视最近的数据
        if (this.neckHistory.length > 0) {
            const weights = this.generateWeights(this.neckHistory.length);
            const smoothNeck = this.weightedAverage(this.neckHistory, weights);
            const smoothTorso = this.weightedAverage(this.torsoHistory, weights);
            
            return {
                neck: smoothNeck,
                torso: smoothTorso
            };
        }
        
        // 初始值
        return {
            neck: neckAngle,
            torso: torsoAngle
        };
    }
    
    reset() {
        this.neckHistory = [];
        this.torsoHistory = [];
    }
}

// ========== Phase 1 优化：滞后阈值评估器（Hysteresis）==========
// 功能：防止在阈值边界频繁切换状态，减少80%的误报
/**
 * 滞后阈值评估器
 * 从"好"切换到"坏"：需要超过阈值 + 滞后值
 * 从"坏"切换到"好"：需要低于阈值 - 滞后值
 */
class HysteresisEvaluator {
    constructor(neckThreshold, torsoThreshold, hysteresis = HYSTERESIS_VALUE) {
        this.neckThreshold = neckThreshold;
        this.torsoThreshold = torsoThreshold;
        this.hysteresis = hysteresis; // 滞后值（度），默认使用 HYSTERESIS_VALUE
        this.currentState = 'good'; // 'good' | 'bad'
    }
    
    /**
     * 评估姿势状态（带滞后）
     * @param {number} neckAngle - 颈部角度
     * @param {number} torsoAngle - 躯干角度
     * @returns {boolean} true=良好姿势, false=不良姿势
     */
    evaluate(neckAngle, torsoAngle) {
        if (this.currentState === 'good') {
            // 从"好"切换到"坏"：需要超过阈值 + 滞后值
            if (neckAngle >= this.neckThreshold + this.hysteresis || 
                torsoAngle >= this.torsoThreshold + this.hysteresis) {
                this.currentState = 'bad';
                return false;
            }
            return true;
        } else {
            // 从"坏"切换到"好"：需要低于阈值 - 滞后值
            if (neckAngle < this.neckThreshold - this.hysteresis && 
                torsoAngle < this.torsoThreshold - this.hysteresis) {
                this.currentState = 'good';
                return true;
            }
            return false;
        }
    }
    
    /**
     * 更新阈值（用于动态调整）
     */
    updateThresholds(neckThreshold, torsoThreshold) {
        this.neckThreshold = neckThreshold;
        this.torsoThreshold = torsoThreshold;
    }
    
    reset() {
        this.currentState = 'good';
    }
}

// ========== Phase 2 优化：自适应阈值管理器（适合SCI患者康复过程）==========
/**
 * 自适应阈值管理器
 * 根据康复进度自动调整阈值
 */
class AdaptiveThresholdManager {
    constructor(baseNeckThreshold = 40, baseTorsoThreshold = 10) {
        this.baseNeckThreshold = baseNeckThreshold;
        this.baseTorsoThreshold = baseTorsoThreshold;
        this.rehabLevel = 'early'; // 'early' | 'middle' | 'late'
        this.multipliers = {
            'early': { neck: 1.5, torso: 2.0 },    // 早期：更宽松（60°, 20°）
            'middle': { neck: 1.2, torso: 1.5 },  // 中期：中等（48°, 15°）
            'late': { neck: 1.0, torso: 1.0 }     // 后期：标准（40°, 15°）
        };
    }
    
    /**
     * 获取当前阈值（根据康复阶段）
     */
    getThresholds() {
        const mult = this.multipliers[this.rehabLevel];
        return {
            neck: this.baseNeckThreshold * mult.neck,
            torso: this.baseTorsoThreshold * mult.torso
        };
    }
    
    /**
     * 根据历史数据自动调整康复阶段
     * @param {Array} historyData - 历史记录数据
     */
    updateRehabLevel(historyData) {
        if (!historyData || historyData.length === 0) {
            this.rehabLevel = 'early';
            return;
        }
        
        // 分析最近7天的数据
        const recentData = historyData.slice(-7);
        const avgGoodPercentage = recentData.reduce((sum, d) => {
            const goodPercent = d.summary ? d.summary.goodPercentage : 
                              (d.goodTime && d.badTime ? 
                               (d.goodTime / (d.goodTime + d.badTime) * 100) : 0);
            return sum + goodPercent;
        }, 0) / recentData.length;
        
        // 根据良好姿势占比判断康复阶段
        if (avgGoodPercentage < 30) {
            this.rehabLevel = 'early';
        } else if (avgGoodPercentage < 70) {
            this.rehabLevel = 'middle';
        } else {
            this.rehabLevel = 'late';
        }
    }
    
    /**
     * 手动设置康复阶段
     */
    setRehabLevel(level) {
        if (['early', 'middle', 'late'].includes(level)) {
            this.rehabLevel = level;
        }
    }
    
    /**
     * 获取康复阶段描述
     */
    getRehabLevelDescription(language = 'zh') {
        const descriptions = {
            zh: {
                early: '早期康复（宽松阈值）',
                middle: '中期康复（中等阈值）',
                late: '后期康复（标准阈值）'
            },
            en: {
                early: 'Early Rehabilitation (Relaxed)',
                middle: 'Middle Rehabilitation (Moderate)',
                late: 'Late Rehabilitation (Standard)'
            }
        };
        return descriptions[language]?.[this.rehabLevel] || this.rehabLevel;
    }
}

// 导出所有类到全局作用域
window.AngleSmoother = AngleSmoother;
window.HysteresisEvaluator = HysteresisEvaluator;
window.AdaptiveThresholdManager = AdaptiveThresholdManager;

File: phase1_validation_tool/algorithms.py
"""
Advanced Posture Detection Algorithms
高级姿态检测算法

Ported from JavaScript version to maintain consistency between Python and Web implementations.
从 JavaScript 版本移植，保持 Python 和 Web 实现的一致性。
"""

import math
from typing import Dict, List, Tuple, Optional


class AngleSmoother:
    """
    Weighted Moving Average Smoother (WMA)
    加权移动平均平滑器
    
    Improvement: More emphasis on recent data, faster response, better smoothing effect
    改进：更重视最近的数据，响应更快，平滑效果更好
    """
    
    def __init__(self, window_size: int = 10):
        """
        Initialize angle smoother
        
        Args:
            window_size: Size of the smoothing window (default: 10 frames)
        """
        self.window_size = window_size
        self.neck_history: List[float] = []
        self.torso_history: List[float] = []
    
    def generate_weights(self, length: int) -> List[float]:
        """
        Generate linear weights (recent data has higher weight)
        生成线性权重（最近的数据权重更高）
        
        Args:
            length: Number of data points
        
        Returns:
            List of normalized weights
        """
        weights = [i + 1 for i in range(length)]  # Weights: 1, 2, 3, ...
        weight_sum = sum(weights)
        return [w / weight_sum for w in weights]  # Normalize
    
    def weighted_average(self, values: List[float], weights: List[float]) -> float:
        """
        Calculate weighted average
        计算加权平均值
        """
        return sum(val * weight for val, weight in zip(values, weights))
    
    def smooth(self, neck_angle: float, torso_angle: float) -> Dict[str, float]:
        """
        Smooth angles using weighted moving average
        使用加权移动平均平滑角度
        
        Args:
            neck_angle: Current neck angle
            torso_angle: Current torso angle
        
        Returns:
            Dict with smoothed 'neck' and 'torso' angles
        """
        # Add new values
        self.neck_history.append(neck_angle)
        self.torso_history.append(torso_angle)
        
        # Maintain window size
        if len(self.neck_history) > self.window_size:
            self.neck_history.pop(0)
            self.torso_history.pop(0)
        
        # Weighted Moving Average (WMA)
        if len(self.neck_history) > 0:
            weights = self.generate_weights(len(self.neck_history))
            smooth_neck = self.weighted_average(self.neck_history, weights)
            smooth_torso = self.weighted_average(self.torso_history, weights)
            
            return {
                'neck': smooth_neck,
                'torso': smooth_torso
            }
        
        # Initial values
        return {
            'neck': neck_angle,
            'torso': torso_angle
        }
    
    def reset(self):
        """Reset history"""
        self.neck_history = []
        self.torso_history = []


class HysteresisEvaluator:
    """
    Hysteresis Threshold Evaluator
    滞后阈值评估器
    
    Function: Prevent frequent state switching at threshold boundaries, reduce 80% false positives
    功能：防止在阈值边界频繁切换状态，减少 80% 的误报
    
    Switching from "good" to "bad": Must exceed threshold + hysteresis value
    从"良好"切换到"不良"：必须超过阈值 + 滞后值
    
    Switching from "bad" to "good": Must be below threshold - hysteresis value
    从"不良"切换到"良好"：必须低于阈值 - 滞后值
    """
    
    def __init__(self, neck_threshold: float, torso_threshold: float, hysteresis: float = 2.0):
        """
        Initialize hysteresis evaluator
        
        Args:
            neck_threshold: Neck angle threshold
            torso_threshold: Torso angle threshold
            hysteresis: Hysteresis value in degrees (default: 2°)
        """
        self.neck_threshold = neck_threshold
        self.torso_threshold = torso_threshold
        self.hysteresis = hysteresis
        self.current_state = 'good'  # 'good' | 'bad'
    
    def evaluate(self, neck_angle: float, torso_angle: float) -> bool:
        """
        Evaluate posture state with hysteresis
        使用滞后评估姿态状态
        
        Args:
            neck_angle: Current neck angle
            torso_angle: Current torso angle
        
        Returns:
            True if good posture, False if bad posture
        """
        if self.current_state == 'good':
            # Switching from "good" to "bad": Must exceed threshold + hysteresis
            if (neck_angle >= self.neck_threshold + self.hysteresis or 
                torso_angle >= self.torso_threshold + self.hysteresis):
                self.current_state = 'bad'
                return False
            return True
        else:
            # Switching from "bad" to "good": Must be below threshold - hysteresis
            if (neck_angle < self.neck_threshold - self.hysteresis and 
                torso_angle < self.torso_threshold - self.hysteresis):
                self.current_state = 'good'
                return True
            return False
    
    def update_thresholds(self, neck_threshold: float, torso_threshold: float):
        """Update thresholds (for dynamic adjustment)"""
        self.neck_threshold = neck_threshold
        self.torso_threshold = torso_threshold
    
    def reset(self):
        """Reset state"""
        self.current_state = 'good'


class AdaptiveThresholdManager:
    """
    Adaptive Threshold Manager
    自适应阈值管理器
    
    Automatically adjusts thresholds based on rehabilitation progress
    根据康复进度自动调整阈值
    
    Suitable for SCI patient rehabilitation process
    适用于 SCI 患者康复过程
    """
    
    def __init__(self, base_neck_threshold: float = 40, base_torso_threshold: float = 15):
        """
        Initialize adaptive threshold manager
        
        Args:
            base_neck_threshold: Base neck angle threshold (default: 40°)
            base_torso_threshold: Base torso angle threshold (default: 15°)
        """
        self.base_neck_threshold = base_neck_threshold
        self.base_torso_threshold = base_torso_threshold
        self.rehab_level = 'early'  # 'early' | 'middle' | 'late'
        
        # Multipliers for different rehabilitation stages
        self.multipliers = {
            'early': {'neck': 1.5, 'torso': 2.0},    # Early: More lenient (60°, 30°)
            'middle': {'neck': 1.2, 'torso': 1.5},   # Middle: Moderate (48°, 22.5°)
            'late': {'neck': 1.0, 'torso': 1.0}      # Late: Standard (40°, 15°)
        }
    
    def get_thresholds(self) -> Dict[str, float]:
        """
        Get current thresholds based on rehabilitation stage
        根据康复阶段获取当前阈值
        
        Returns:
            Dict with 'neck' and 'torso' thresholds
        """
        mult = self.multipliers[self.rehab_level]
        return {
            'neck': self.base_neck_threshold * mult['neck'],
            'torso': self.base_torso_threshold * mult['torso']
        }
    
    def update_rehab_level(self, history_data: List[Dict]) -> None:
        """
        Automatically adjust rehabilitation stage based on historical data
        根据历史数据自动调整康复阶段
        
        Args:
            history_data: List of historical records
        """
        if not history_data or len(history_data) == 0:
            self.rehab_level = 'early'
            return
        
        # Analyze data from the last 7 days
        recent_data = history_data[-7:]
        
        # Calculate average good posture percentage
        total_percentage = 0
        for record in recent_data:
            if 'summary' in record and 'goodPercentage' in record['summary']:
                good_percent = record['summary']['goodPercentage']
            elif 'goodTime' in record and 'badTime' in record:
                total_time = record['goodTime'] + record['badTime']
                good_percent = (record['goodTime'] / total_time * 100) if total_time > 0 else 0
            else:
                good_percent = 0
            
            total_percentage += good_percent
        
        avg_good_percentage = total_percentage / len(recent_data)
        
        # Determine rehabilitation stage based on good posture percentage
        if avg_good_percentage < 30:
            self.rehab_level = 'early'
        elif avg_good_percentage < 70:
            self.rehab_level = 'middle'
        else:
            self.rehab_level = 'late'
    
    def set_rehab_level(self, level: str) -> None:
        """
        Manually set rehabilitation stage
        手动设置康复阶段
        
        Args:
            level: 'early', 'middle', or 'late'
        """
        if level in ['early', 'middle', 'late']:
            self.rehab_level = level
    
    def get_rehab_level_description(self, language: str = 'en') -> str:
        """
        Get rehabilitation stage description
        获取康复阶段描述
        
        Args:
            language: 'en' or 'zh'
        
        Returns:
            Description string
        """
        descriptions = {
            'zh': {
                'early': '早期康复（宽松阈值）',
                'middle': '中期康复（中等阈值）',
                'late': '后期康复（标准阈值）'
            },
            'en': {
                'early': 'Early Rehabilitation (Relaxed)',
                'middle': 'Middle Rehabilitation (Moderate)',
                'late': 'Late Rehabilitation (Standard)'
            }
        }
        
        return descriptions.get(language, descriptions['en']).get(self.rehab_level, self.rehab_level)


def calculate_angle_precise(p1: Dict[str, float], p2: Dict[str, float], p3: Dict[str, float]) -> float:
    """
    High-precision three-point angle calculation (vector dot product method)
    高精度三点角度计算（向量点积法）
    
    Precision improvement: From ±3° to ±0.5°, suitable for medical applications
    精度提升：从 ±3° 提升至 ±0.5°，适合医学应用
    
    Args:
        p1: Reference point (shoulder/hip) {'x': float, 'y': float}
        p2: Target point (ear/shoulder) {'x': float, 'y': float}
        p3: Vertical reference point (above p1) {'x': float, 'y': float}
    
    Returns:
        Angle in degrees, precision ±0.5°
    """
    try:
        # Vector 1: p1 -> p2
        v1 = {
            'x': p2['x'] - p1['x'],
            'y': p2['y'] - p1['y']
        }
        
        # Vector 2: p1 -> p3 (vertical reference vector)
        v2 = {
            'x': p3['x'] - p1['x'],
            'y': p3['y'] - p1['y']
        }
        
        # Calculate vector magnitudes
        mag1 = math.sqrt(v1['x'] ** 2 + v1['y'] ** 2)
        mag2 = math.sqrt(v2['x'] ** 2 + v2['y'] ** 2)
        
        # Check validity (avoid division by zero)
        if mag1 < 1e-6 or mag2 < 1e-6:
            return 0
        
        # Calculate dot product
        dot = v1['x'] * v2['x'] + v1['y'] * v2['y']
        
        # Calculate angle (radians)
        cos_angle = dot / (mag1 * mag2)
        clamped_cos = max(-1, min(1, cos_angle))
        angle_rad = math.acos(clamped_cos)
        
        # Convert to degrees (keep 1 decimal place)
        return round((180 / math.pi) * angle_rad, 1)
    
    except Exception as e:
        print(f"Warning: High-precision angle calculation failed - {e}")
        return 0


def median(arr: List[float]) -> float:
    """
    Calculate median (resistant to outliers)
    计算中位数（抗离群值）
    
    Args:
        arr: List of values
    
    Returns:
        Median value
    """
    if len(arr) == 0:
        return 0
    
    # Filter out invalid values
    sorted_arr = sorted([v for v in arr if not math.isnan(v) and v > 0])
    
    if len(sorted_arr) == 0:
        return 0
    
    mid = len(sorted_arr) // 2
    
    if len(sorted_arr) % 2 == 0:
        return (sorted_arr[mid - 1] + sorted_arr[mid]) / 2
    else:
        return sorted_arr[mid]


def calculate_angle_with_fusion(
    landmarks: List,
    canvas_width: int,
    canvas_height: int,
    angle_type: str,
    landmark_indices: Dict[str, int]
) -> float:
    """
    Multi-keypoint fusion angle calculation (improve robustness)
    多关键点融合角度计算（提高鲁棒性）
    
    Args:
        landmarks: MediaPipe landmarks array
        canvas_width: Canvas width
        canvas_height: Canvas height
        angle_type: 'neck' or 'torso'
        landmark_indices: Dictionary of landmark indices
    
    Returns:
        Fused angle value (median of multiple calculations)
    """
    angles = []
    
    try:
        if angle_type == 'neck':
            # Method 1: Left shoulder - Left ear
            l_shoulder = landmarks[landmark_indices['LEFT_SHOULDER']]
            l_ear = landmarks[landmark_indices['LEFT_EAR']]
            
            if l_shoulder and l_ear:
                shoulder = {'x': l_shoulder.x * canvas_width, 'y': l_shoulder.y * canvas_height}
                ear = {'x': l_ear.x * canvas_width, 'y': l_ear.y * canvas_height}
                ref = {'x': shoulder['x'], 'y': shoulder['y'] - 100}
                
                angle = calculate_angle_precise(shoulder, ear, ref)
                if angle > 0:
                    angles.append(angle)
            
            # Method 2: Right shoulder - Right ear (if visible)
            r_shoulder = landmarks[landmark_indices['RIGHT_SHOULDER']]
            r_ear = landmarks[landmark_indices['RIGHT_EAR']]
            
            if r_shoulder and r_ear:
                shoulder = {'x': r_shoulder.x * canvas_width, 'y': r_shoulder.y * canvas_height}
                ear = {'x': r_ear.x * canvas_width, 'y': r_ear.y * canvas_height}
                ref = {'x': shoulder['x'], 'y': shoulder['y'] - 100}
                
                angle = calculate_angle_precise(shoulder, ear, ref)
                if angle > 0:
                    angles.append(angle)
        
        elif angle_type == 'torso':
            # Method 1: Left hip - Left shoulder
            l_hip = landmarks[landmark_indices['LEFT_HIP']]
            l_shoulder = landmarks[landmark_indices['LEFT_SHOULDER']]
            
            if l_hip and l_shoulder:
                hip = {'x': l_hip.x * canvas_width, 'y': l_hip.y * canvas_height}
                shoulder = {'x': l_shoulder.x * canvas_width, 'y': l_shoulder.y * canvas_height}
                
                if hip['y'] > shoulder['y']:
                    ref = {'x': hip['x'], 'y': hip['y'] - 100}
                    angle = calculate_angle_precise(hip, shoulder, ref)
                    if angle > 0:
                        angles.append(angle)
            
            # Method 2: Right hip - Right shoulder (if visible)
            r_hip = landmarks[landmark_indices['RIGHT_HIP']]
            r_shoulder = landmarks[landmark_indices['RIGHT_SHOULDER']]
            
            if r_hip and r_shoulder:
                hip = {'x': r_hip.x * canvas_width, 'y': r_hip.y * canvas_height}
                shoulder = {'x': r_shoulder.x * canvas_width, 'y': r_shoulder.y * canvas_height}
                
                if hip['y'] > shoulder['y']:
                    ref = {'x': hip['x'], 'y': hip['y'] - 100}
                    angle = calculate_angle_precise(hip, shoulder, ref)
                    if angle > 0:
                        angles.append(angle)
    
    except Exception as e:
        print(f"Warning: Angle fusion calculation failed - {e}")
    
    # Return median (resistant to outliers)
    return median(angles)


# Export all classes and functions
__all__ = [
    'AngleSmoother',
    'HysteresisEvaluator',
    'AdaptiveThresholdManager',
    'calculate_angle_precise',
    'median',
    'calculate_angle_with_fusion'
]

